### PR TITLE
Improve E0211 error diagnostic

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @Manishearth

cc #29248

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -2348,8 +2348,8 @@ For information on the design of the orphan rules, see [RFC 1023].
 "##,
 
 E0211: r##"
-You used an intrinsic function which doesn't correspond to its
-definition. Erroneous code example:
+You used a function/type which doesn't correspond to its
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: "You used a function or type which doesn't fit the requirements for where it was used"

(definition is wrong here)

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -2357,15 +2357,71 @@ definition. Erroneous code example:
...
 ```
 #![feature(intrinsics)]
 
 extern "rust-intrinsic" {
-    fn size_of<T>() -> usize;
+    fn size_of<T>() -> usize; // ok!
+}
+```
+
+The second case example is a bit particular : start functions must always
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: `main` functions. `start()` actually does take parameters

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -2357,15 +2357,71 @@ definition. Erroneous code example:
...
+```
+let x = 1u8;
+match x {
+    0u8...3u8 => (), // ok!
+    _ => ()
+}
+```
+
+And finally, for the last example, only `Box<Self>`, `&Self`, `Self`,
+or `&mut Self` work as explicit self params. Example:
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: *parameters

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: I thought it was an invalid type/function which didn't correspond to what was expected. Whar is still wrong ?

Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: Not all of these have a corresponding "definition"

Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Which one doesn't have then ? And what "global" explanation would you give instead ?

File: src/librustc_typeck/diagnostics.rs [hunk 4]
```diff
@@ -2357,15 +2357,71 @@ definition. Erroneous code example:
...
 ```
 #![feature(intrinsics)]
 
 extern "rust-intrinsic" {
-    fn size_of<T>() -> usize;
+    fn size_of<T>() -> usize; // ok!
+}
+```
+
+The second case example is a bit particular : main function must always
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: the main function

**Comment by GuillaumeGomez** [comment]
Updated !

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -2357,15 +2357,71 @@ definition. Erroneous code example:
...
 ```
 #![feature(intrinsics)]
 
 extern "rust-intrinsic" {
-    fn size_of<T>() -> usize;
+    fn size_of<T>() -> usize; // ok!
+}
+```
+
+The second case example is a bit particular : the main function must always
```
Comment (durka) [inline src/librustc_typeck/diagnostics.rs]: odd spacing here with the colon

Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Oups, french habit...

**Pull request merged**